### PR TITLE
#155 Better warnings (UI) and notifications (in App) 

### PR DESCRIPTION
**Pull request by ArdJonker**
Specifically for errors regarding no_schedules available.
Persistent notifications for configuration error.

I know it has become (a little) big, maybe it's best to look at this together..

### CODE REVIEW

File: wallbox_client.py [hunk 1]
```diff
@@ -70,10 +74,10 @@ def get_charger_state(self) -> int:
...
             if attempts > max_attempts:
                 # Assume the charger has crashed.
-                self.log(f"get_charger_state has reached max attempts ({attempts}), the charger probably crashed. "
-                         f"Set charge_mode to Stop, notify user and try restarting charger.")
-                self.notify_user("Critical error. Automatic charging has been stopped. "
-                                 "Please open the V2G Liberty App to solve this problem.", critical=True)
+                self.log(f"get_charger_state has reached max attempts ({attempts}), the charger probably crashed. ")
+                title = "Critical error"
+                message = "Automatic charging has been stopped. Please click this notification to open the V2G Liberty App and follow the steps to solve this problem."
+                self.notify_user(message, title, "critical_error", True, False)
```
Comment (ArdJonker) [inline wallbox_client.py:80]: Consider removing after charger restart  and switching back to automatic.

File: wallbox_client.py [hunk 2]
```diff
@@ -169,6 +173,10 @@ def set_charger_action(self, action: str):
             # To counter this we call "stop" from disconnect event.
             value = self.registers["actions"]["stop_charging"]
         elif action == "restart":
+            # AJO 2023-11-30
+            # This seems irellevant code as the restart is (only) needed when the modbus module
+            # of the charger has crashed and then it does not receive this instruction anymore.
+            # Remove?
```
Comment (ArdJonker) [inline wallbox_client.py:179]: Checken

File: wallbox_client.py [hunk 3]
```diff
@@ -419,10 +427,8 @@ def process_soc(self, reported_soc: str) -> bool:
 
         # Notify user of reaching 80% charge while charging (not dis-charging).
         # ToDo: Discuss with users if this is useful.
-        # ToDo: Replace 80 with max soc setting from yaml.
-        notify_soc_setting = 80
-        if self.connected_car_soc == notify_soc_setting and self.is_charging():
-            self.notify_user(f"Car battery state of charge has reached {notify_soc_setting}%.")
+        if self.connected_car_soc == c.CAR_MAX_SOC_IN_PERCENT and self.is_charging():
+            self.notify_user(f"Car battery at {self.connected_car_soc}%, range ≈ {tmp}km.", None, "battery_max_soc_reached", 60*15)
```
Comment (ArdJonker) [inline wallbox_client.py:431]: add space before km

File: wallbox_client.py (hunk 2 above)
Comment (ArdJonker) [reply wallbox_client.py:179]: Will do this later.

File: wallbox_client.py (hunk 3 above)
Comment (ArdJonker) [reply wallbox_client.py:431]: According to international standards (for technical documentation) there should be a space between number and UoM. To make sure they are not separated over two lines a non-breaking narrow " " space is used.

File: v2g_liberty.py [hunk 4]
```diff
@@ -74,13 +84,22 @@ def initialize(self):
...
+        self.no_schedule_notification_is_planned = False
+
         self.client = self.configure_charger_client()
         self.log_errors()
         self.get_app("flexmeasures-client").authenticate_with_fm()
 
         self.listen_state(self.update_charge_mode, "input_select.charge_mode", attribute="all")
         self.listen_state(self.handle_charger_state_change, "sensor.charger_charger_state", attribute="all")
-        self.listen_event(self.restart_charger, "RESTART_CHARGER")
+        self.listen_event(self.run_test_code, "RESTART_CHARGER")
```
Comment (Flix6x) [inline v2g_liberty.py:102]: This is now running test code?

File: v2g_liberty.py [hunk 5]
```diff
@@ -119,44 +184,160 @@ def disconnect_charger(self, *args, **kwargs):
...
+                self.no_schedule_notification_is_planned = True
         else:
-            self.notify(message, title=title)
+            self.set_state("input_boolean.error_no_new_schedule_available", state="off")
+            canceled_before_run = self.cancel_timer(self.notification_timer_handle)
+            if self.no_schedule_notification_is_planned and not canceled_before_run:
+                # Only send this message if the previous wa
+                title = "Scheduals available again"
+                message = f"The problems with schedules have been solved. " \
+                          f"If you've set charging via the chargers app, consider end that and use automatic charging agian."
```
Comment (Flix6x) [inline v2g_liberty.py:331]: consider end -> consider to end
agian -> again

File: wallbox_client.py [hunk 6]
```diff
@@ -15,6 +15,10 @@ class WallboxModbusMixin:
     NUM_MODBUS_PORTS = 65536
     last_restart: int = 0
 
+    # This is the what the charger returns for a disconnected state
```
Comment (Flix6x) [inline wallbox_client.py:18]: Still unclear if 0 means disconnected and 1 means connected, or vice versa.

**Review by Flix6x — CHANGES_REQUESTED** [review]

**Comment by Flix6x** [comment]
It looks like my last three review comments are still unresolved.

File: v2g_liberty.py (hunk 4 above)
Comment (ArdJonker) [reply v2g_liberty.py:102]: Remove

File: v2g_liberty.py (hunk 5 above)
Comment (ArdJonker) [reply v2g_liberty.py:319]: OK

File: wallbox_client.py (hunk 6 above)
Comment (ArdJonker) [reply wallbox_client.py:18]: OK, will explain

Comment (Flix6x) [reply wallbox_client.py]: To be honest it is still unclear what a 0 state means and what a 1 state means.

File: v2g_liberty.py (hunk 4 above)
Comment (Flix6x) [reply v2g_liberty.py]: The line originally called `self.restart_charger`. That's no longer needed?

File: v2g_liberty.py (hunk 5 above)
Comment (Flix6x) [reply v2g_liberty.py]: Looks like you skipped my second suggestion.

File: v2g_liberty.py (hunk 4 above)
Comment (ArdJonker) [reply v2g_liberty.py]: Indeed, this code did not really serve a purpose. Restarting the charger is needed when the communication via modbus with the charger is not possible. As this restart function is using modbus it was not useful. I'll also remove the restart button from the dashboard.

File: wallbox_client.py (hunk 6 above)
Comment (ArdJonker) [reply wallbox_client.py]: I'll try to add a more clear explanation.

**Review by Flix6x — APPROVED** [review]
Thanks also for the more thorough typo corrections.

**Pull request merged**